File: algorithm/206.cpp
#include <vector>
#include <cstddef>
using namespace std;

struct ListNode
{
	int val;
	ListNode* next;
	ListNode(int x): val(x), next(NULL) {};
};

class Solution
{
public:
	ListNode* reverseList(ListNode* head);
};

ListNode* Solution::reverseList(ListNode* head)
{
	if(head==NULL || head->next==NULL) return head;
	ListNode* cur = head;
	vector<ListNode*> tmp = {};
	
	while(cur!=NULL)
	{
		tmp.push_back(cur);
		cur=cur->next;
	}
	
	ListNode* head2 = tmp.back();
	cur = head2;
	tmp.pop_back();
	

	while(!tmp.empty())
	{
		ListNode* node = tmp.back();
		tmp.pop_back();
		cur->next = node;
		node->next = NULL;
		cur=cur->next;
	}
	return head2;
}

int main()
{
	ListNode* n1 = new ListNode(1);
	ListNode* n2 = new ListNode(2);
	ListNode* n3 = new ListNode(3);
	ListNode* n4 = new ListNode(4);
	ListNode* n5 = new ListNode(5);

	n1->next = n2;
	n2->next = n3;
	n3->next = n4;
	n4->next = n5;

	Solution obj;
	ListNode* result = obj.reverseList(n1);
	return 0;

}

File: algorithm/go/firstround/206.go
package firstround

type ListNode struct {
	Val  int
	Next *ListNode
}

// 递归，时间复杂度O(N)，空间复杂度O(N)
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	// last: 表示反转从head.Next到链表末尾的链表，反转后链表的头节点，即原链表的最后一个节点
	last := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return last
}

// 迭代，时间复杂度O(N)，空间复杂度O(1)
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	cur := head
	var prev *ListNode
	for cur != nil {
		next := cur.Next
		cur.Next = prev
		prev = cur
		cur = next
	}
	return prev
}
